### ISSUE TITLE
[15.3/sdk2] User can't update FSharp.Core/System.ValueTuple using the nuget GUI

### ISSUE THREAD

**Issue by 0x53A** [Microsoft/visualfsharp#3361]
Start with the simplest sdk project:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="File1.fs" />
  </ItemGroup>

</Project>
```

Now open the nuget gui:

__1) the versions shown to the user are wrong__

![image](https://user-images.githubusercontent.com/4236651/28470022-971de698-6e38-11e7-91a0-3ba65a3e1c6c.png)

I then select System.ValueTuple and click on update:

![image](https://user-images.githubusercontent.com/4236651/28470064-c45c898e-6e38-11e7-9cb7-9e4522240a10.png)

Looks good!

This is how the project file looks:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="File1.fs" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="System.ValueTuple" Version="4.4.0-preview2-25405-01" />
  </ItemGroup>

</Project>

```

After a project reload, the gui looks like this:

![image](https://user-images.githubusercontent.com/4236651/28470126-fe318998-6e38-11e7-86ea-fda162095bb0.png)

So,

__2) Nuget update does not work, but the gui pretends it does until the next reload__

### Cause

Nuget adds ``<PackageReference Update``, because it thinks the project already contains the package.

But the package is only added later, so the ``update`` is a no-op.

I think it would work if it instead added ``<PackageReference Include``.

**Comment by forki** [comment]
This should be reported to dotnet repos. It's not a vf# bug

**Comment by 0x53A** [comment]
Yes it is, because nuget obviously does not know about ```DisableImplicitFSharpCoreReference``` and ```DisableImplicitSystemValueTupleReference```.

**Comment by forki** [comment]
It's also general nuget issue of which thing has priority. User specified or whatever the sdk ships

**Comment by 0x53A** [comment]
Yes, but in this case I would discriminate between sdk and nuget. Nuget in this case adds the ``<PackageReference``, which is correct, it is the SDK that ignores it.

**Comment by 0x53A** [comment]
I just want to call attention that this extra-sausage of the F#-sdk will cause issues with the rest of the tooling. 

Different example:

Base project:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="File1.fs" />
  </ItemGroup>

</Project>
```

I googled how to update a package via the CLI and found https://ardalis.com/how-do-i-update-a-nuget-package-using-dotnet-cli

`` dotnet add package System.ValueTuple -v 4.4.0-preview2-25405-01``

```
  Writing C:\Users\lrieger\AppData\Local\Temp\tmpA84F.tmp
info : Adding PackageReference for package 'System.ValueTuple' into project 'Test.fsproj'.
log  : Restoring packages for Test.fsproj...
info :   CACHE https://api.nuget.org/v3-flatcontainer/fsharp.core/index.json
info : Package 'System.ValueTuple' is compatible with all the specified frameworks in project 'Test.fsproj'.
error: Error while performing Update for package 'System.ValueTuple'. Cannot edit items in imported files -
error:   Item 'PackageReference' for 'System.ValueTuple' in Imported file 'C:\Program Files\dotnet\sdk\2.1.0-preview1-006784\FSharp\Microsoft.FSharp.NetSdk.targets'.

```

Is this a nuget issue? Maybe. But it works with a C# project.

**Comment by forki** [comment]
All true. Just wished someone would finally  report the general case to
nuget/dotnet. Ours are all special cases of the underlying problem

Am 21.07.2017 17:36 schrieb "Lukas Rieger" <notifications@github.com>:

**Comment by 0x53A** [comment]
@forki these two implicit references are imported here:

https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Build/Microsoft.FSharp.NetSdk.targets#L72

So this is (imo) plainly an issue with the F# sdk.

**Comment by forki** [comment]
> , the last one wins 

Did you test that?

**Comment by 0x53A** [comment]
As far as I know the targets file get's imported at the end of the project:

```Xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="File1.fs" />
  </ItemGroup>

<!-- Microsoft.FSharp.NetSdk.targets is implicitly referenced here -->
</Project>
```

This is just by convention, props are imported at the top, targets at the end.

**Comment by forki** [comment]
Ok if they get imported at the end then they always win. And user settings get overriden. I personally think this is a serious issue

**Comment by 0x53A** [comment]
Wait, let me test that again, maybe I'm talking bullshit. Nuget added ``<PackageReference Update``, which is different to ``<PackageReference Include``.

**Comment by 0x53A** [comment]
Now I'm totally confused, it does look like the first one wins:

```
    <PackageReference Include="Newtonsoft.Json" Version="10.0.1" />
    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
```

==> V10

----------


```
    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
    <PackageReference Include="Newtonsoft.Json" Version="10.0.1" />
```

==> V9

----

**Comment by 0x53A** [comment]
Wait, the first one wins, and the sdk is imported last.

This file

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.ValueTuple" Version="4.4.0-preview1-25305-02" />
    <PackageReference Include="Newtonsoft.Json" Version="10.0.1" />-->
    <PackageReference Include="FSharp.Core" Version="4.0.0.1" />
  </ItemGroup>

</Project>
```

produces this nupkg:

![image](https://user-images.githubusercontent.com/4236651/28476026-c5c58620-6e4e-11e7-9057-05de89cb0569.png)

**Comment by 0x53A** [comment]
So for some reason (and that the gui in visual studio shows _something_ , which may be different from the truth probably does not help) I thought that you could not change the version of FSharp.Core without setting ``DisableImplicitFSharpCoreReference``. I will test it again, but from the previous comment above I conclude that this is not true; if the user already imported any other FSharp.Core, then the include by the sdk is a no-op.

If that is correct, then everything is great again.

**Comment by forki** [comment]
Apparently this happens with 4.5 as well: https://twitter.com/CarstenK_Dev/status/1009119383640268801?s=19

**Comment by cartermp** [comment]
Not sure what the root cause is (sounds like it's not a VF# issue), but F# 4.5 and associated bits are not released in any form on the SDK yet (that's 2.1.400). We don't have a guarantee of forwards-compatibility, so I'm not entirely unsurprised that this doesn't work.
